carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,282 @@
+
+# import graph building methods
+from sklearn.neighbors import kneighbors_graph, radius_neighbors_graph
+
+
+def graph_search(
+    data,
+    index,
+    keys_immutable,
+    model,
+    n_neighbors=50,
+    p_norm=2,
+    mode="knn",
+    frac=0.4,
+    radius=0.25,
+):
+    # This one implements the FACE method from
+    # Rafael Poyiadzi et al (2020), "FACE: Feasible and Actionable Counterfactual Explanations",
+    # Conference on AI, Ethics & Accountability (AIES, 2020)
+    """
+    :param data: df
+    :param n_neighbors: int > 0; number of neighbors when constructing knn graph
+    :param step: float > 0; step_size for growing spheres
+    :param mode: str; either 'knn' or 'epsilon'
+    :param model: classification model (either tf keras, pytorch or sklearn)
+    :param p_norm: float=>1; denotes the norm (classical: 1 or 2)
+    :param frac: float 0 < number =< 1; fraction of data for which we compute the graph; if frac = 1, and data set large, then compute long
+    :param keys_immutable: list; list of input names that may not be searched over
+    :param radius: float > 0; parameter for epsilon density graph
+    :return: candidate_counterfactual_star: np array (min. cost counterfactual explanation)
+    """
+    # Choose a subset of data for computational efficiency
+    data = choose_random_subset(data, frac, index)
+
+    # ADD CONSTRAINTS by immutable inputs into adjacency matrix
+    # if element in adjacency matrix 0, then it cannot be reached
+    # this ensures that paths only take same sex / same race / ... etc. routes
+    # TODO: CANNOT deal with continuous constraints (e.g. age) yet
+    # TODO: However, authors also do not mention how to go about it
+    for i in range(len(keys_immutable)):
+        immutable_constraint_matrix1, immutable_constraint_matrix2 = build_constraints(
+            data, i, keys_immutable
+        )
+
+    # POSITIVE PREDICTIONS
+    y_predicted = model.predict_proba(data.values)
+    y_predicted = np.argmax(y_predicted, axis=1)
+    y_positive_indeces = np.where(y_predicted == 1)
+
+    # TODO: replace counterfactual search by function, which takes 'mode' as an argument
+    candidate_counterfactuals_star = []
+    if mode == "knn":
+        boundary = 3  # chosen in ad-hoc fashion
+        median = n_neighbors
+        is_knn = True
+
+    elif mode == "epsilon":
+        boundary = 0.10  # chosen in ad-hoc fashion
+        median = radius
+        is_knn = False
+    else:
+        raise ValueError("Only possible values for mode are knn and epsilon")
+
+    neighbors_list = [
+        median - boundary,
+        median,
+        median + boundary,
+    ]
+
+    # obtain candidate targets (CT); two conditions need to be met:
+    # (1) CT needs to be predicted positively & (2) CT needs to have certain "density"
+    # for knn I interpret 'certain density' as sufficient number of neighbours
+    for n in neighbors_list:
+        find_cf(
+            candidate_counterfactuals_star,
```

Comment:
Lets do a more functional programming syle here and assign it to a return variable instead of passing it as a parameter and mutating it from the inside of the function. (harder to read and more prone to side-effects)

```
candidate_counterfactuals = []
for n in neighbors_list:
    neighbor_candidates = find_cf(...)
    candidate_counterfactuals.append(neighbor_candidates)
```